nixos/modules/services/networking/vpp.nix

```diff
@@ -0,0 +1,543 @@
+          #   }
+          #   foo bar2 {
+          #     baz2
+          #   }
+          set = concatMap (
+            subname:
+            lib.optionals (value.${subname} != null) (
+              [ "${name} ${subname} {" ] ++ (map (line: "  ${line}") (toLines value.${subname})) ++ [ "}" ]
+            )
+          ) (filter (v: v != null) (builtins.attrNames value));
+        }
+        .${builtins.typeOf value};
+
+      # One level "above" encode, acts upon a set and uses encode on each name,value pair
+      toLines = set: concatMap (name: encode name set.${name}) (sortedAttrs set);
+
+      # Moves top-level attrsets into a dummy "" value, so that `section = {...}` is transformed to `section  {...}`
+      parseSections =
+        set:
+        builtins.mapAttrs (name: value: {
+          "" = value;
+        }) set;
+    in
+    lib.strings.concatStringsSep "\n" (toLines (parseSections cfg));
+
+  semanticTypes = with types; rec {
+    vppAtom = nullOr (oneOf [
+      int
+      bool
+      str
+    ]);
+    vppAttr = attrsOf vppAll;
+    vppAll =
+      (oneOf [
+        vppAtom
+        (listOf vppAtom)
+        vppAttr
+      ])
+      // {
+        # Since this is a recursive type and the description by default contains
+        # the description of its subtypes, infinite recursion would occur without
+        # explicitly breaking this cycle
+        description = "vpp values (atoms (null, str, int, bool), list of atoms, or attrsets of vpp values)";
+      };
+  };
+
+  vppOpts =
+    { name, ... }:
+    {
+      options = {
+        enable = mkEnableOption "FD.io's Vector Packet Processor, a high-performance userspace network stack.";
+
+        name = mkOption {
+          type = types.str;
+          default = name;
+          description = ''
+            Name is used as a suffix for the service name.
+            By default it takes the value you use for `<instance>` in:
+            {option}`services.vpp.instances.<instance>`
+          '';
+        };
+
+        package = mkPackageOption pkgs "vpp" { };
+
+        kernelModule = mkOption {
+          type = types.nullOr types.str;
+          default = "uio_pci_generic";
+          example = "vfio-pci";
+          description = "UIO kernel driver module to load before starting this VPP instance. Set to `null` to disable this functionality.";
+        };
+
+        group = mkOption {
+          type = types.str;
+          default = "vpp";
+          example = "vpp-main";
+          description = "Group that grants users in it privileges to control this instance via {command}`vppctl`.";
+        };
+
+        settings = mkOption {
+          description = ''
+            Configuration for VPP, see <https://fd.io/docs/vpp/master/configuration/reference.html> for details.
+
+            Nix value declared here will be translated directly to the config format VPP uses. Attributes
+            called `extraConfig` will be inserted verbatim into the resulting config file.
```

Comment:
The parser was mostly copied from [znc](https://github.com/NixOS/nixpkgs/blob/c636130b369ff944bb1c3560c8bb9377e6082975/nixos/modules/services/networking/znc/default.nix)'s module along with its `extraConfig` impl, and I think now that it's there there's no point in going back to a classical implementation. IMO being able to insert text into otherwise nix-expressed sections is pretty idiomatic, especially since the parser makes it possible to express `extraConfig` as a list of strings instead of one big multiline string.

Reply:
If it's necessary for certain configs to be possible (i.e specifying an attribute twice may override one of the definitions) due to the way `vpp` configuration syntax is I might think differently, but I'm inclined to think that `znc` is the outlier here - the code in question hasn't changed in six years, and it still uses the now discouraged `with lib;`. Again, I'm not an authoritative source, but I'm not sure that I see the benefit here.